### PR TITLE
[WB-8039] Accept categorical probabilities in grid search 

### PR DESCRIPTION
**Pull request by bcsherma**
https://wandb.atlassian.net/browse/WB-8039

This PR introduces the option to explicitly specify the distribution for categorical random variables. In the definition of a categorical parameter within the sweep config, the `probabilities` key can be added with an array of numbers; the array must be in equal in length to the cardinality of the variable and the values in the array must sum to 1 (each value must be >= 0).

The choice to make the array sum to 1 is not necessary, we could instead allow the user to provide an array of positive numbers and then normalize it ourselves.

Some constraints on the probability array (e.g. summing to 1, equal length to # values) are not expressible via the schema, so new checks have been added to cfg.py

This PR adds tests both for the validation of the new sweep config features and testing that the samples from custom categorical random variables match the spec. The sampling test is implemented by checking that number of samples drawn for a particular value is in the range of x +- 2(x^0.5) where x is the expected number of samples for that particular value per the parameter config. This can be modified to be more stringent, but the number of samples may need to increase for the test to not become overly flaky.

### CODE REVIEW

**Review by dannygoldstein — COMMENTED** [review]
instead of augmenting the existing `param_categorical` json subschema lets make a separate subschema called `param_categorical_w_probabilities` where probabilities are a required parameter and the other parameters are the same as the standard categorical. the reason for this is that i want to leave the existing categorical path, which is pretty hot, as is and let this touch separate code.

**Review by dannygoldstein — COMMENTED** [review]
looks good, let's just add that one piece in about disallowing categorical probabilities on grid search (see final comment above)

File: tests/test_config.py [hunk 1]
```diff
@@ -47,3 +47,23 @@ def test_missing_parameters_section():
...
+        "method": "random",
+        "parameters": {
+            "v1": {"values": [1, 2, 3], "probabilities": [0.1, 0.2, 0.3, 0.4]}
+        },
+    }
+    with pytest.raises(jsonschema.ValidationError):
+        _ = config.SweepConfig(invalid_config)
+
+
+def test_irregular_probs():
```
[the remark this answers is not in the document]
Comment (bcsherma) [reply tests/test_config.py:63]: I think this checking is happening automatically when we use a separate subschema for the categorical variables with params through [this path](https://github.com/wandb/sweeps/blob/222e41b91667cdbc1dd9df30371da80f432df93c/src/sweeps/grid_search.py#L57). Should I just

[the remark this answers is not in the document]
Comment (bcsherma) [reply tests/test_config.py:63]: I can add a test case though

[the remark this answers is not in the document]
Comment (dannygoldstein) [reply tests/test_config.py:63]: Ah good point, and thanks that would be great

[the remark this answers is not in the document]
Comment (dannygoldstein) [reply tests/test_config.py:63]: Can we update the error message to make clear that probabilities cannot be specified for grid search parameters

[the remark this answers is not in the document]
Comment (bcsherma) [reply tests/test_config.py:63]: Definitely

[the remark this answers is not in the document]
Comment (dannygoldstein) [reply tests/test_config.py:63]:
```python
                f"Parameter {p.name} is a disallowed type with grid search. Grid search requires all parameters "
                f"to be categorical, constant, int_uniform, or q_uniform. Specification of probabilities for categorical parameters is disallowed in grid search"
```

or something

**Review by dannygoldstein — APPROVED** [review]

**Comment by dannygoldstein** [comment]
@bcsherma is there a jira ticket we should link to this?

**Pull request merged**